sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ResolvedDataSource.scala

```diff
@@ -130,7 +141,49 @@ object ResolvedDataSource extends Logging {
       bucketSpec: Option[BucketSpec],
       provider: String,
       options: Map[String, String]): ResolvedDataSource = {
-    val clazz: Class[_] = lookupDataSource(provider)
+    // Here, it tries to find out data source by file extensions if the `format()` is not called.
+    // The auto-detection is based on given paths and it recognizes glob pattern as well but
+    // it does not recursively check the sub-paths even if the given paths are directories.
+    // This source detection goes the following steps
+    //
+    //   1. Check `provider` and use this if this is not `null`.
+    //   2. If `provider` is not given, then it tries to detect the source types by extension.
+    //      at this point, if detects only if all the given paths have the same extension.
+    //   3. if it fails to detect, use the datasource given to `spark.sql.sources.default`.
+    //
+    val paths = {
```

Comment:
note that i'd move this detection code into a separate class, so we can unit test it.